init/devices.cpp

```diff
@@ -399,44 +411,112 @@ std::vector<std::string> DeviceHandler::GetBlockDeviceSymlinks(const Uevent& uev
+
+        MakeDevice(dev_path, block, major, minor, all_links);
+
+        for (const auto& link_path : all_links) {
+            std::string link_target;
+
+            if (!mkdir_recursive(Dirname(link_path), 0755))
+                PLOG(ERROR) << "Failed to create directory " << Dirname(link_path);
+
+            // Only create a link in /dev/aliases for the device
+            // aliases we created from an 'alias' line in the
+            // configuration. This is a way to save what aliases we've
+            // created in /dev for a specific device and allow use to
+            // find the exact aliases we created without keeping state
+            // within ueventd.
+
+            // alias_link is the link (eg: /dev/modem -> /dev/ttyACM0)
+            // that we want to be creating.
+            if (!alias_link_path.empty()
+                && link_path == alias_link_path
+                && symlink(alias_link_path.c_str(), state_link_path.c_str())) {
+
+                // If creating the symlink failed, it can be because
+                // of a misc error, or it can be because the file
+                // already exists. We don't want to replace the file
+                // so we make it an explicit error message that the
+                // configuration is wrong.
+
+                if (errno != EEXIST)
+                    PLOG(ERROR) << "Failed to create alias symlink from " << state_link_path
+                                << " to " << alias_link_path;
+
+                else if (Readlink(state_link_path, &link_target) && link_target != alias_link_path)
+                    PLOG(ERROR) << "Failed to create alias symlink from " << state_link_path
+                                << " to " << alias_link_path << ", which already links to: " << link_target;
             }
-
-            if (symlink(devpath.c_str(), link.c_str())) {
-                if (errno != EEXIST) {
-                    PLOG(ERROR) << "Failed to symlink " << devpath << " to " << link;
-                } else if (std::string link_path;
-                           Readlink(link, &link_path) && link_path != devpath) {
-                    PLOG(ERROR) << "Failed to symlink " << devpath << " to " << link
-                                << ", which already links to: " << link_path;
+            else {
+                if (symlink(dev_path.c_str(), link_path.c_str())) {
+                    if (errno != EEXIST)
+                        PLOG(ERROR) << "Failed to symlink " << dev_path
+                                    << " to " << link_path;
+
+                    else if (Readlink(link_path, &link_target) && link_target != dev_path)
+                        PLOG(ERROR) << "Failed to symlink " << dev_path
+                                    << " to " << link_path << ", which already links to: " << link_target;
+
+                    // Delete the link in /dev/aliases
+                    unlink(alias_link_path.c_str());
```

Comment:
Don't you mean `state_link_path`? The one that's a path under `/dev/alias`

Reply:
Yes you're right